Read schema defaults when generating field using EntityContext

ADmad opened the pull request:
In 2.x when creating an input if field value was not set in request data the default value available from table schema was set as field value (if an explicit default value was not provided).

This patch implements the same feature.

Refs #8733, Closes #5065

ADmad committed:
Read schema defaults when generating field using EntityContext.

ADmad committed:
Fix issue when providing explicit default value for input.

markstory commented on src/View/Form/EntityContext.php:
```diff
@@ -230,6 +251,26 @@ public function val($field)
+        if ($table === false) {
+            return null;
+        }
+        $defaults = $table->schema()->defaultValues();
```
What do you think about having a method to fetch the default value of a field? `$schema->defaultValue($column);` as an example.

ADmad replied:
I was waiting for the `Schema/Table` class name changes to land. After that I can add a new method and update this method to use `defultValue($columns)` instead of `defaultValues()` later on.

markstory replied:
Ok.

markstory commented:
Looking good so far.

ionas commented:
Thanks!
p.s.: Just out of curiosity - does it work for foreign keys and belongsTo (default selects).

ADmad commented:
@markstory I have added the formhelper integrations tests. Anything more is needed?

@ionas It will work for any field which has a column default set in table.

ADmad committed:
Update signature of `ContextInterface::val()`

All classes implementing ContextInterface now take into account
`$options` passes to `val()` method.

ADmad commented:
I have added `$options` argument to `ContextInterface::val()` and updated all implementing classes.

ADmad committed:
Update signature of NullContext::val().

markstory commented:
A bunch of builds are failing on both travis and appveyor. Might be related.

ADmad commented:
Yup I have checked and ensured there are no failures due to my changes. The failure in `Cake\Test\TestCase\View\Form\EntityContextTest::testValOnCollections()` is due to bug in `SqlSchema` for which PR is open.

lorenzo commented on src/View/Form/EntityContext.php:
```diff
@@ -220,7 +230,18 @@ public function val($field)
+                return $val;
+            }
+            if ($options['default'] !== null
+                || !$options['schemaDefault']
```
If I pass a default in options, I want that to take precedence over the schema

ADmad replied:
It does. Check [this](https://github.com/cakephp/cakephp/pull/8735/files#diff-cf0be6f0051f81c7f12aaa4c6becafc7R4053) test.

ADmad commented:
Ping

lorenzo commented:
Good job @ADmad, but it seems like your code does not pass tests in appveyor. I restarted the build just in case, bit the failures do seem relevant

ADmad commented:
@lorenzo Those failures were caused due to a bug in `SqliteSchema` which were fixed in #8734. I merged `master` into `3.next` and rebased my branch on top, so not sure why those failures are still showing up on appveyor when there are no failures on travis.

lorenzo commented:
@ADmad but appveyor does not even has pdo_sqlite installed

ADmad commented:
Ah right it uses Sqlserver :)  Then it seems `SqlserverSchema` needs the same fix that was done for `SqliteSchema`.

ADmad commented:
@lorenzo I have opened #8769 which has test showing `SqlserverSchema` has problem with `DEFAULT NULL` similar to one that `SqliteSchema` had.

ADmad commented:
The issue with `SqlserverSchema` has been fixed in #8769.

This PR is ready to be merged.

markstory commented on src/View/Form/ContextInterface.php:
```diff
@@ -48,9 +48,14 @@ public function isCreate();
      * @return mixed
      */
-    public function val($field);
+    public function val($field, $options = []);
```
We'll need to update the elasticsearch plugin too. This is an API breaking change unfortunately :cry: as changing the interface will cause fatal errors in userland code.

ADmad replied:
PHP doesn't seem to complain if the `$options` param is removed from interface method. So can we have it in implementing classes only?  The classes would have to be updated to account for the options but at least there won't be any signature mismatch errors until they are updated.

ADmad replied:
Line [2416](https://github.com/cakephp/cakephp/pull/8735/files#diff-86d5537968f770710b98fddb2840fcdaR2416) in `FormHelper` will ensure that setting `default` key still works as it currently does, even if a context class ignores `$options`.

markstory replied:
Excluding the argument from the interface, documenting the second argument, and only including the argument in the relevant implementations is a good compromise :+1:

ADmad replied:
Cool, I'll update accordingly.

markstory committed:
Fix typo in documentation.

ADmad committed:
Remove $options argument from ContextInterface::val() to maintain BC.

ADmad commented:
@markstory Updates done.

markstory commented:
This looks good to me. :+1:
